File: python_backend/utils/crypto_envelope.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import os
from typing import Any, Callable, Dict, Optional

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from ..services import get_config

ENVELOPE_VERSION = 1
ENVELOPE_ALGORITHM = "aes-256-gcm"


def _stable_json(value: Any) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def _derive_key(secret: str) -> bytes:
    raw = str(secret or "").strip()
    if not raw:
        raise RuntimeError("DATA_ENCRYPTION_KEY is required for encrypted data access")
    try:
        decoded = base64.b64decode(raw, validate=True)
        if len(decoded) == 32:
            return decoded
    except Exception:
        pass
    return hashlib.sha256(raw.encode("utf-8")).digest()


def _master_key() -> bytes:
    config = get_config()
    key = str((config.encryption or {}).get("key") or "").strip()
    return _derive_key(key)


def _blind_index_key() -> bytes:
    config = get_config()
    configured = str((config.encryption or {}).get("blind_index_key") or "").strip()
    if configured:
        return _derive_key(configured)
    return _master_key()


def _key_version() -> str:
    config = get_config()
    return str((config.encryption or {}).get("key_version") or "local-v1").strip() or "local-v1"


def _kms_key_id() -> Optional[str]:
    config = get_config()
    value = str((config.encryption or {}).get("kms_key_id") or "").strip()
    return value or None


def _canonical_aad(aad: Optional[Dict[str, Any]]) -> bytes:
    if not aad:
        return b""
    return _stable_json(aad).encode("utf-8")


def _wrap_data_key(data_key: bytes) -> Dict[str, str]:
    wrapper = AESGCM(_master_key())
    iv = os.urandom(12)
    aad = _canonical_aad(
        {
            "purpose": "wrapped_data_key",
            "key_version": _key_version(),
            "kms_key_id": _kms_key_id(),
        }
    )
    wrapped = wrapper.encrypt(iv, data_key, aad)
    return {
        "alg": ENVELOPE_ALGORITHM,
        "iv": base64.b64encode(iv).decode("ascii"),
        "ciphertext": base64.b64encode(wrapped).decode("ascii"),
    }


def _unwrap_data_key(wrapped_data_key: Any) -> bytes:
    if not isinstance(wrapped_data_key, dict):
        raise ValueError("wrapped_data_key must be an object")
    iv = base64.b64decode(str(wrapped_data_key.get("iv") or ""))
    ciphertext = base64.b64decode(str(wrapped_data_key.get("ciphertext") or ""))
    wrapper = AESGCM(_master_key())
    aad = _canonical_aad(
        {
            "purpose": "wrapped_data_key",
            "key_version": _key_version(),
            "kms_key_id": _kms_key_id(),
        }
    )
    return wrapper.decrypt(iv, ciphertext, aad)


def compute_blind_index(
    value: Any,
    *,
    label: str,
    normalizer: Optional[Callable[[str], str]] = None,
) -> Optional[str]:
    if value is None:
        return None
    text = str(value).strip()
    if not text:
        return None
    normalized = normalizer(text) if callable(normalizer) else text
    digest = hmac.new(
        _blind_index_key(),
        f"{label}:{normalized}".encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()
    return digest


def encrypt_text(
    value: Any,
    *,
    aad: Optional[Dict[str, Any]] = None,
    blind_index: Optional[str] = None,
) -> Optional[str]:
    if value is None:
        return None
    text = str(value)
    if not text:
        return None
    data_key = os.urandom(32)
    cipher = AESGCM(data_key)
    iv = os.urandom(12)
    ciphertext = cipher.encrypt(iv, text.encode("utf-8"), _canonical_aad(aad))
    envelope = {
        "version": ENVELOPE_VERSION,
        "alg": ENVELOPE_ALGORITHM,
        "kms_key_id": _kms_key_id(),
        "key_version": _key_version(),
        "wrapped_data_key": _wrap_data_key(data_key),
        "iv": base64.b64encode(iv).decode("ascii"),
        "ciphertext": base64.b64encode(ciphertext).decode("ascii"),
        "aad": aad or {},
    }
    if blind_index:
        envelope["blind_index"] = blind_index
    return json.dumps(envelope, separators=(",", ":"), ensure_ascii=False)


def decrypt_text(value: Any, *, aad: Optional[Dict[str, Any]] = None) -> Optional[str]:
    if value is None:
        return None
    if isinstance(value, (bytes, bytearray)):
        try:
            value = value.decode("utf-8")
        except Exception:
            return None
    if not isinstance(value, str):
        return None
    text = value.strip()
    if not text:
        return None
    try:
        payload = json.loads(text)
    except Exception:
        return text
    if not isinstance(payload, dict):
        return text

    if payload.get("version") == ENVELOPE_VERSION and payload.get("wrapped_data_key"):
        data_key = _unwrap_data_key(payload.get("wrapped_data_key"))
        cipher = AESGCM(data_key)
        iv = base64.b64decode(str(payload.get("iv") or ""))
        ciphertext = base64.b64decode(str(payload.get("ciphertext") or ""))
        plaintext = cipher.decrypt(iv, ciphertext, _canonical_aad(aad or payload.get("aad") or {}))
        return plaintext.decode("utf-8")

    if "iv" in payload and "payload" in payload:
        legacy_key = _master_key()
        cipher = AESGCM(legacy_key)
        iv = base64.b64decode(payload["iv"])
        ciphertext = base64.b64decode(payload["payload"])
        plaintext = cipher.decrypt(iv, ciphertext, None)
        return plaintext.decode("utf-8")

    return text


def encrypt_json(
    value: Any,
    *,
    aad: Optional[Dict[str, Any]] = None,
    blind_index: Optional[str] = None,
) -> Optional[str]:
    if value is None:
        return None
    return encrypt_text(_stable_json(value), aad=aad, blind_index=blind_index)


def decrypt_json(value: Any, *, aad: Optional[Dict[str, Any]] = None) -> Any:
    text = decrypt_text(value, aad=aad)
    if text is None:
        return None
    try:
        return json.loads(text)
    except Exception:
        return None

File: server/utils/cryptoEnvelope.js
const crypto = require('crypto');
const { env } = require('../config/env');

const ENVELOPE_VERSION = 1;
const ENVELOPE_ALGORITHM = 'aes-256-gcm';

const stableSort = (value) => {
  if (Array.isArray(value)) {
    return value.map((item) => stableSort(item));
  }
  if (value && typeof value === 'object') {
    return Object.keys(value)
      .sort()
      .reduce((acc, key) => {
        acc[key] = stableSort(value[key]);
        return acc;
      }, {});
  }
  return value;
};

const stableJson = (value) => JSON.stringify(stableSort(value));

const deriveKey = (secret) => {
  const raw = String(secret || '').trim();
  if (!raw) {
    throw new Error('DATA_ENCRYPTION_KEY is required for encrypted data access');
  }
  try {
    const decoded = Buffer.from(raw, 'base64');
    if (decoded.length === 32 && decoded.toString('base64').replace(/=+$/, '') === raw.replace(/=+$/, '')) {
      return decoded;
    }
  } catch (_error) {
    // ignore and fall back to a deterministic digest of the configured secret
  }
  return crypto.createHash('sha256').update(raw, 'utf8').digest();
};

const getMasterKey = () => deriveKey(env.encryption?.key);
const getBlindIndexKey = () => deriveKey(env.encryption?.blindIndexKey || env.encryption?.key);
const getKeyVersion = () => String(env.encryption?.keyVersion || 'local-v1').trim() || 'local-v1';
const getKmsKeyId = () => {
  const value = String(env.encryption?.kmsKeyId || '').trim();
  return value || null;
};

const canonicalAad = (aad) => {
  if (!aad || typeof aad !== 'object') {
    return Buffer.alloc(0);
  }
  return Buffer.from(stableJson(aad), 'utf8');
};

const wrapDataKey = (dataKey) => {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', getMasterKey(), iv);
  const aad = canonicalAad({
    purpose: 'wrapped_data_key',
    key_version: getKeyVersion(),
    kms_key_id: getKmsKeyId(),
  });
  cipher.setAAD(aad);
  const ciphertext = Buffer.concat([cipher.update(dataKey), cipher.final(), cipher.getAuthTag()]);
  return {
    alg: ENVELOPE_ALGORITHM,
    iv: iv.toString('base64'),
    ciphertext: ciphertext.toString('base64'),
  };
};

const unwrapDataKey = (wrappedDataKey) => {
  if (!wrappedDataKey || typeof wrappedDataKey !== 'object') {
    throw new Error('wrapped_data_key must be an object');
  }
  const iv = Buffer.from(String(wrappedDataKey.iv || ''), 'base64');
  const ciphertext = Buffer.from(String(wrappedDataKey.ciphertext || ''), 'base64');
  const tag = ciphertext.subarray(ciphertext.length - 16);
  const payload = ciphertext.subarray(0, ciphertext.length - 16);
  const decipher = crypto.createDecipheriv('aes-256-gcm', getMasterKey(), iv);
  const aad = canonicalAad({
    purpose: 'wrapped_data_key',
    key_version: getKeyVersion(),
    kms_key_id: getKmsKeyId(),
  });
  decipher.setAAD(aad);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(payload), decipher.final()]);
};

const computeBlindIndex = (value, { label, normalizer } = {}) => {
  const text = value === null || value === undefined ? '' : String(value).trim();
  if (!text) return null;
  const normalized = typeof normalizer === 'function' ? normalizer(text) : text;
  return crypto
    .createHmac('sha256', getBlindIndexKey())
    .update(`${label || 'default'}:${normalized}`, 'utf8')
    .digest('hex');
};

const encryptText = (value, { aad, blindIndex } = {}) => {
  const text = value === null || value === undefined ? '' : String(value);
  if (!text) return null;
  const dataKey = crypto.randomBytes(32);
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', dataKey, iv);
  const aadBuffer = canonicalAad(aad);
  if (aadBuffer.length) {
    cipher.setAAD(aadBuffer);
  }
  const ciphertext = Buffer.concat([cipher.update(text, 'utf8'), cipher.final(), cipher.getAuthTag()]);
  const envelope = {
    version: ENVELOPE_VERSION,
    alg: ENVELOPE_ALGORITHM,
    kms_key_id: getKmsKeyId(),
    key_version: getKeyVersion(),
    wrapped_data_key: wrapDataKey(dataKey),
    iv: iv.toString('base64'),
    ciphertext: ciphertext.toString('base64'),
    aad: aad && typeof aad === 'object' ? stableSort(aad) : {},
  };
  if (blindIndex) {
    envelope.blind_index = blindIndex;
  }
  return JSON.stringify(envelope);
};

const decryptText = (value, { aad } = {}) => {
  if (value === null || value === undefined) return null;
  const text = Buffer.isBuffer(value) ? value.toString('utf8') : String(value).trim();
  if (!text) return null;
  let payload;
  try {
    payload = JSON.parse(text);
  } catch (_error) {
    return text;
  }
  if (!payload || typeof payload !== 'object') {
    return text;
  }

  if (payload.version === ENVELOPE_VERSION && payload.wrapped_data_key) {
    const dataKey = unwrapDataKey(payload.wrapped_data_key);
    const iv = Buffer.from(String(payload.iv || ''), 'base64');
    const ciphertext = Buffer.from(String(payload.ciphertext || ''), 'base64');
    const tag = ciphertext.subarray(ciphertext.length - 16);
    const encrypted = ciphertext.subarray(0, ciphertext.length - 16);
    const decipher = crypto.createDecipheriv('aes-256-gcm', dataKey, iv);
    const aadBuffer = canonicalAad(aad || payload.aad || {});
    if (aadBuffer.length) {
      decipher.setAAD(aadBuffer);
    }
    decipher.setAuthTag(tag);
    return Buffer.concat([decipher.update(encrypted), decipher.final()]).toString('utf8');
  }

  if (payload.iv && payload.payload) {
    const iv = Buffer.from(String(payload.iv || ''), 'base64');
    const ciphertext = Buffer.from(String(payload.payload || ''), 'base64');
    const tag = ciphertext.subarray(ciphertext.length - 16);
    const encrypted = ciphertext.subarray(0, ciphertext.length - 16);
    const decipher = crypto.createDecipheriv('aes-256-gcm', getMasterKey(), iv);
    decipher.setAuthTag(tag);
    return Buffer.concat([decipher.update(encrypted), decipher.final()]).toString('utf8');
  }

  return text;
};

const encryptJson = (value, options = {}) => {
  if (value === null || value === undefined) return null;
  return encryptText(stableJson(value), options);
};

const decryptJson = (value, options = {}) => {
  const text = decryptText(value, options);
  if (!text) return null;
  try {
    return JSON.parse(text);
  } catch (_error) {
    return null;
  }
};

module.exports = {
  ENVELOPE_ALGORITHM,
  ENVELOPE_VERSION,
  computeBlindIndex,
  decryptJson,
  decryptText,
  encryptJson,
  encryptText,
};
